app/services/listing_index_service/api/listings.rb

```diff
@@ -0,0 +1,111 @@
+module ListingIndexService::API
+
+  SelectionGroups = EntityUtils.define_builder(
+    [:values, :array, :mandatory],
+    [:search_type, one_of: [:and, :or]],
+  )
+
+  NumericFilter = EntityUtils.define_builder(
+    [:id, :fixnum, :mandatory],
+    [:range, :range, :mandatory],
+  )
+
+  SearchParams = EntityUtils.define_builder(
+    [:page, :to_integer, default: 1, gte: 1],
+    [:per_page, :to_integer, :mandatory, gte: 1],
+    [:keywords, :string, :optional],
+    [:category_id, :fixnum, :optional],
+    [:listing_shape_id, :fixnum, :optional],
+    [:price_cents, :range, :optional],
+    [:checkboxes, entity: SelectionGroups],
+    [:dropdowns, entity: SelectionGroups],
```

Comment:
Checkboxes and dropdown menus are representation details. The data type needs to be changed.